File: Assignment-3/3/quick_sort3.py
from random import randint
"""
    ****************DIVIDE AND CONQUER****************
    * Take last element as pivot
    * Compare the pivot element with all other elements in the sub array
    * If the element is equal to pivot element then decrement
      pivot index and swap the traversing index and pivot index
    * If the element is smaller increase k-index
    * Else dont increment it.
    => Finally k-index will be the starting index of the elements
       bigger than pivot element and all the elements equal to pivot element 
       will be in the last
    => So shift the last portion with k portion
    => Then the pivot and all of its redundant elements will be in final position 
       of the array 
    => Continue this to get Sorted array
"""
def swap(array, i, j):
    temp = array[i]
    array[i] = array[j]
    array[j] = temp

def minimum(a, b):
    if a < b:
        return a
    return b

def quick_sort(array, low, high):
    
    if low > high:
        return
    pivot_index = randint(low , high)
    swap(array, high, pivot_index)
    
    i = low
    k = low
    p = high
    while i < p:
        if array[i] < array[p]:
            swap(array, i , k)
            i += 1
            k += 1
        elif array[i] == array[p]:
            p -= 1
            swap(array, i, p)
        else:
            i += 1
    
    m = minimum(p - k, high - p + 1)

    for i in range(0, m):
        swap(array, k + i, high - m + 1 + i)
    quick_sort(array, low, k - 1)
    quick_sort(array, high - p + k + 1, high)

def main():

    #n = int(input("Give n: "))
    n = int(input())
    print("Give n elements: ")
    array = [int(x) for x in input().split()]
    assert(len(array) == n)
    quick_sort(array, 0 , n - 1)
    print("Sorted array: ")
    print(array)

if __name__ == "__main__":
    main()

File: Assignment-3/3/quick_sort3.cpp
#include<iostream>
#include<vector>

/*
    ****************DIVIDE AND CONQUER****************
    * Take last element as pivot
    * Compare the pivot element with all other elements in the sub array
    * If the element is equal to pivot element then decrement
      pivot index and swap the traversing index and pivot index
    * If the element is smaller increase k-index
    * Else dont increment it.
    => Finally k-index will be the starting index of the elements
       bigger than pivot element and all the elements equal to pivot element 
       will be in the last
    => So shift the last portion with k portion
    => Then the pivot and all of its redundant elements will be in final position 
       of the array 
    => Continue this to get Sorted array
*/
using namespace std;

void swap(int &a, int &b)
{
    if(a == b)
        return;
    int temp = a;
    a = b;
    b = temp;
}

int min(int a, int b)
{
    if(a < b)
        return a;
    return b;
}

void quick_sort3(vector<long long> &list, int low, int high)
{
    if(low > high)
        return;
    int pivot_index = low + rand() % ((high + 1) - low);
    swap(list[high], list[pivot_index]);
    int i = low;
    int k = low;
    int pivot = high;

    while(i < pivot)
    {
        if(list[i] < list[pivot])
        {
            swap(list[i], list[k]);
            i++;
            k++;
        }
        else if(list[i] == list[pivot])
        {
            pivot--;
            swap(list[i], list[pivot]);
        }
        else
        {
            i++;
        }
    }
    int m = min(pivot - k, high - pivot + 1);
    
    for(int lc = 0; lc < m; lc++)
    {
        swap(list[k + lc], list[high - m + 1 + lc]);
    }
    quick_sort3(list, low, k - 1);
    quick_sort3(list, high - pivot + k + 1, high);
}

int main()
{

    int n = 0;
    //cout<<"Give n: ";
    cin>>n;
    
    vector<long long> list(n);

    //cout<<"Give unsorted elements: ";
    for(int i = 0; i < n; i++)
        cin>>list[i];


    quick_sort3(list, 0, n - 1);
    //cout<<"Sorted elements are :";
    for(int i = 0; i < n; i++)
        cout<<list[i]<<" ";
    cout<<"\n";
    return 0;
}
